### ISSUE TITLE
redbean listing page shouldn't show subdirectories

### ISSUE THREAD

**Issue by coderofsalvation** [jart/cosmopolitan#681]
Is it possible / is there a flag for redbean to list all directories & contents? Basically flip hide/show-directories-by-default?

```
$ redbean.com -D . &
```

> this only lists the root dir..not the subdirectories from the disk

In the docs I read: 

```The listing page only applies to the root directory. However the default index page applies to subdirectories too. In order for it to work, there needs to be an empty directory entry in the zip. That should already be the default practice of your zip editor.```

I don't fully understand this..So i tried:

```
$ mkdir foo
$ zip redbean.com foo
$ mkdir foo/bar
$ echo helloworld foo/bar/hello.txt
$ redbean.com -D . &
```

> I was hoping that it would somehow override it, but I didn't get the open directory

Is there a simple way around this (preferrably without too much lines of lua) 

## Background

Rclone allows easy mounting of open http directories (example is [modland.com](https://modland.com))

```
$ apt-get install rclone    # change this to your distro install cmd
$ rclone config
n 
name> modland.com
type of storage> 33
url> https://modland.com
Edit advanced config? (y/n) y
headers> <enter>
no_head> false
q

$ rclone mount modland:/pub/modules ~/modland.com --daemon
$ ls ~/modland.com
PROFIT!
```

**Comment by jart** [comment]
It took me a while to parse this because I hadn't heard of rclone. Thank you for explaining it though. It seems if I use it like SSHFS to mount the redbean-demo.com HTTP server, then it gets confused by the fact that we just show the entire directory tree hierarchy on one page.

```
master jart@nightmare:~/vendor/libc-test$ ls -1 /mnt/redbean/
404.html
binarytrees.lua
call-lua-module.lua
closedsource.lua
favicon.ico
fetch.lua
finger.lua
gensvg.lua
help.txt
hitcounter.lua
maxmind.lua
opensource.lua
redbean.lua
redbean.png
seekable.txt
sql-backup.lua
sql.lua
statusz
store-asset.lua
tool
unix-dir.lua
unix-finger.lua
unix-info.lua
unix-rawsocket.lua
unix-subprocess.lua
unix-unix.lua
unix-webserver.lua
master jart@nightmare:~/vendor/libc-test$ ls -1 /mnt/redbean/tool
ls: reading directory '/mnt/redbean/tool': I/O error
```

Whereas our listing says:

```
/.cosmo                                 2022-03-16T00:00:00  100644  n/a        0  
/.init.lua                              2022-03-16T00:00:00  100644  55%    2,565  
/.lua/                                  2022-03-16T00:00:00  040755  n/a        0  
/.lua/mymodule.lua                      2022-03-16T00:00:00  100644  14%      104  
/.reload.lua                            2022-03-16T00:00:00  100644  n/a       63  
[/404.html ](http://10.10.10.124:8080/404.html)                              2022-03-16T00:00:00  100644  57%      436  
[/binarytrees.lua ](http://10.10.10.124:8080/binarytrees.lua)                       2022-03-16T00:00:00  100644  59%    6,077  
[/call-lua-module.lua ](http://10.10.10.124:8080/call-lua-module.lua)                   2022-03-16T00:00:00  100644  58%    1,558  
[/closedsource.lua ](http://10.10.10.124:8080/closedsource.lua)                      2022-03-16T00:00:00  100644  12%      113  
[/crashreport.lua ](http://10.10.10.124:8080/crashreport.lua)                       2022-03-16T00:00:00  100644  39%      422  
[/favicon.ico ](http://10.10.10.124:8080/favicon.ico)                           2022-03-16T00:00:00  100644  64%   16,958  
[/fetch.lua ](http://10.10.10.124:8080/fetch.lua)                             2022-03-16T00:00:00  100644  59%    2,406  
[/finger.lua ](http://10.10.10.124:8080/finger.lua)                            2022-03-16T00:00:00  100644  60%    1,848  
[/gensvg.lua ](http://10.10.10.124:8080/gensvg.lua)                            2022-03-16T00:00:00  100644  57%    3,014  
[/help.txt ](http://10.10.10.124:8080/help.txt)                              2022-03-16T00:00:00  100644  70%  229,082  
[/hitcounter.lua ](http://10.10.10.124:8080/hitcounter.lua)                        2022-03-16T00:00:00  100644  41%      845  
[/maxmind.lua ](http://10.10.10.124:8080/maxmind.lua)                           2022-03-16T00:00:00  100644  59%   11,096  
[/opensource.lua ](http://10.10.10.124:8080/opensource.lua)                        2022-03-16T00:00:00  100644  16%      102  
[/printpayload.lua ](http://10.10.10.124:8080/printpayload.lua)                      2022-03-16T00:00:00  100644  36%      282  
[/redbean-form.lua ](http://10.10.10.124:8080/redbean-form.lua)                      2022-03-16T00:00:00  100644  66%    1,683  
[/redbean-xhr.lua ](http://10.10.10.124:8080/redbean-xhr.lua)                       2022-03-16T00:00:00  100644  34%      192  
[/redbean.justine.lol/virtualbean.html ](http://10.10.10.124:8080/redbean.justine.lol/virtualbean.html)  2022-03-16T00:00:00  100644  38%      254  
[/redbean.lua ](http://10.10.10.124:8080/redbean.lua)                           2022-03-16T00:00:00  100644  73%   11,486  
[/redbean.png ](http://10.10.10.124:8080/redbean.png)                           2022-03-16T00:00:00  100644  n/a    5,073  
[/seekable.txt ](http://10.10.10.124:8080/seekable.txt)                          2022-03-16T00:00:00  100644  n/a       52  
[/sql-backup.lua ](http://10.10.10.124:8080/sql-backup.lua)                        2022-03-16T00:00:00  100644  35%      391  
[/sql-backupstore.lua ](http://10.10.10.124:8080/sql-backupstore.lua)                   2022-03-16T00:00:00  100644  22%      125  
[/sql.lua ](http://10.10.10.124:8080/sql.lua)                               2022-03-16T00:00:00  100644  11%      133  
[/store-asset.lua ](http://10.10.10.124:8080/store-asset.lua)                       2022-03-16T00:00:00  100644  53%    2,228  
[/tool/ ](http://10.10.10.124:8080/tool/)                                 2022-03-16T00:00:00  040755  n/a        0  
[/tool/net/ ](http://10.10.10.124:8080/tool/net/)                             2022-03-16T00:00:00  040755  n/a        0  
[/tool/net/demo/ ](http://10.10.10.124:8080/tool/net/demo/)                        2022-03-16T00:00:00  040755  n/a        0  
[/tool/net/demo/index.html ](http://10.10.10.124:8080/tool/net/demo/index.html)              2022-03-16T00:00:00  100644  43%      897  
[/tool/net/demo/redbean.css ](http://10.10.10.124:8080/tool/net/demo/redbean.css)             2022-03-16T00:00:00  100644  50%      426  
[/unix-dir.lua ](http://10.10.10.124:8080/unix-dir.lua)                          2022-03-16T00:00:00  100644  72%    2,905  
[/unix-finger.lua ](http://10.10.10.124:8080/unix-finger.lua)                       2022-03-16T00:00:00  100644  57%    2,313  
[/unix-info.lua ](http://10.10.10.124:8080/unix-info.lua)                         2022-03-16T00:00:00  100644  87%    9,486  
[/unix-rawsocket.lua ](http://10.10.10.124:8080/unix-rawsocket.lua)                    2022-03-16T00:00:00  100644  69%    3,843  
[/unix-subprocess.lua ](http://10.10.10.124:8080/unix-subprocess.lua)                   2022-03-16T00:00:00  100644  63%    1,652  
[/unix-unix.lua ](http://10.10.10.124:8080/unix-unix.lua)                         2022-03-16T00:00:00  100644  57%    2,684  
[/unix-webserver.lua ](http://10.10.10.124:8080/unix-webserver.lua)                    2022-03-16T00:00:00  100644  71%    4,104  
/.symtab                                2022-11-06T01:17:48  100644  68%  327,680
```

That's a problem with rclone. It's not a problem with redbean.

You have the ability to address this. You can use redbean's [unix.opendir()](https://redbean.dev/#unix.opendir) API to whip up a some Lua code that causes redbean to list directories the same way as NGINX, Apache, etc. so that it'll just work as an rcloud filesystem. Here's our demo example code which shows how to use unix.opendir():

- https://github.com/jart/cosmopolitan/blob/master/tool/net/demo/unix-dir.lua

If you built a traditional directory listing app for redbean, then we'd love to put it on our list of community modules on both https://redbean.dev and https://github.com/shmup/awesome-cosmopolitan I think rcloud is really cool (first I've heard of it) so it'd be great to have a solution integrating it so that redbean can act as a local file system.

**Comment by coderofsalvation** [comment]
thanks for this.
Btw. I've noticed how confusing my post was (it was late) and I even realized I wrote `$ echo helloworld foo/bar/hello.txt` instead of `$ echo helloworld > foo/bar/hello.txt` :D <br>

Cool, ok in that case I'll whip up the necessary lua-code.
Indeed, it's quite powerful to just mount redbean as a clouddrive. I'm curious whether rclone also supports POST-requests (probably gotta put that behind a http basicauth though)
